plugins/coverage/token.js

```diff
@@ -82,50 +84,57 @@ module.exports = config => ({
 
                 const selfSonar = new CoveragePlugin(selfSonarConfig);
                 const data = await selfSonar.coveragePlugin.getAccessToken(tokenConfig);
-                const projectUrl = selfSonar.coveragePlugin.getProjectData(tokenConfig);
-
-                if (pipeline && projectUrl) {
-                    try {
-                        const pipelineSonarBadge = {
-                            defaultName: pipelineId,
-                            defaultUri: projectUrl
-                        };
-                        let shouldPipelineUpdate = true;
-
-                        if (
-                            pipeline.badges &&
-                            pipeline.badges.sonar &&
-                            pipeline.badges.sonar.defaultName === pipelineId &&
-                            pipeline.badges.sonar.defaultUri === projectUrl
-                        ) {
-                            shouldPipelineUpdate = false;
-                        }
 
-                        if (shouldPipelineUpdate) {
-                            if (pipeline.badges) {
-                                pipeline.badges.sonar = pipelineSonarBadge;
-                            } else {
-                                pipeline.badges = {
-                                    sonar: pipelineSonarBadge
-                                };
-                            }
-
-                            await pipeline.update();
-                            logger.info(
-                                `update pipeline:${pipeline.id}'s sonar badge with defaultName:${pipelineId}, defaultUri: ${projectUrl}`
-                            );
+                return h.response(data);
+            }
+
+            const data = await config.coveragePlugin.getAccessToken(tokenConfig);
+            const { projectUrl } = config.coveragePlugin.getProjectData(tokenConfig);
+
+            if (!pipeline && pipelineId) {
+                pipeline = await pipelineFactory.get(pipelineId);
+
+                logger.info(`looking up again, pipeline:${pipelineId}, and found pipeline: ${pipeline}`);
+            }
+
+            if (pipeline && projectUrl) {
+                try {
+                    const pipelineSonarBadge = {
+                        defaultName: `${pipelineId}`, // ensure pipelineId is stored as String instead of Integer
+                        defaultUri: projectUrl
+                    };
+                    let shouldPipelineUpdate = true;
+
+                    if (
```

Comment:
can we combine this if branch with one at the bottom?

Reply:
Can you give me one example? This is what I can come up with, only update if truly needed.

Reply:
Look in PR https://github.com/screwdriver-cd/screwdriver/pull/2947, I have made multiple attempts to optimize regards to if logic.

![image](https://github.com/screwdriver-cd/screwdriver/assets/15989893/a5d27599-9fbf-413f-93d3-d116ff1cd421)

Reply:
If there's a better way, I will make another PR to optimize further